Enable MultiNamespace and KubeObject protection in DRPC

raghavendra-talur opened the pull request:
This PR adds the functionality to DRPC to protect one or more namespace instead of protecting an application/subscription.

BenamarMk commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,147 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
+func (r *DRPlacementControlReconciler) getProtectedNamespaces(drpc *rmn.DRPlacementControl,
+	log logr.Logger,
+) ([]string, error) {
+	if drpc.Spec.ProtectedNamespaces != nil {
```
Again, `ProtectedNamespaces` could be there but empty
`drpc.Spec.ProtectedNamespaces == &[]`

raghavendra-talur replied:
done

BenamarMk commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,147 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
+	common, err := r.drpcHaveCommonClusters(ctx, drpc, otherDRPC, log)
+	if err != nil {
+		return fmt.Errorf("failed to check if drpcs have common clusters (%w)", err)
+	} else if !common {
```
What if we have the following:
`policy1 {c1, c2}`
`policy2 {c2, c3}`
`common` will be true with `c2`
We wouldn't even get here if we check for this kind of conflict in the drPolicy. I looked quickly but didn't find any. if we don't check, then this is a problem and the fix should probably go in the DRPolicy controller.

BenamarMk reviewed (commented):
@raghavendra-talur finished reviewing drplacementcontrol. I made few comments.
Next (later) I'll review the rest.

BenamarMk commented on controllers/util/drpolicy_util.go:
```diff
@@ -22,6 +22,10 @@ func DrpolicyClusterNames(drpolicy *rmn.DRPolicy) []string {
 	return drpolicy.Spec.DRClusters
 }
 
+func DrpolicyClusterNamesAsASet(drpolicy *rmn.DRPolicy) sets.String {
```
(nit) we should change he name of the function to:
`DRPolicyClusterNamesAsASet`. Same for `DRPolicyClusterNames`
The first 3 letters capitalized.

raghavendra-talur replied:
done

BenamarMk commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -1284,3 +1288,101 @@ func removeString(values []string, s string) []string {
 	return result
 }
+
+func vrgInAdminNamespace(vrg *ramendrv1alpha1.VolumeReplicationGroup, ramenConfig *ramendrv1alpha1.RamenConfig) bool {
```
Duplicate functions. Not really needed. I would create one and call it `IsInAdminNamespace`. I would change the signature to: `(obj  client.Object, ramenConfig *ramendrv1alpha1.RamenConfig)`

raghavendra-talur replied:
I made it two different ones because they have a slightly different behavior. For the VRG, ramen operator namespace is also an admin namespace whereas for the drpc it is not.

Wanted to highlight that in the code.

BenamarMk replied:
Well, I am a bit confused. For `Unmanaged` apps, the VRG namespace is always the same as the DRPC namespace.  If not, then where in the DRPC code that shows the difference?  I am having trouble spotting it.

BenamarMk replied:
My confusion stems from the following comment found in ramenconfig_types.go
```
// RamenOpsNamespace is the namespace where resources for unmanaged apps are created
```
In other words, the `vrgAdminNamespaceNames` is not only checking the `RamenOperandsNamespace` but also the RamenOperatorsNamespace`. If I am confused now, I will be confused when I start debugging this. :)

raghavendra-talur replied:
In the scenario where we have the unmanaged app using the DRPC, the VRG can only be in the same namespace as the DRPC which is the RamenOperandsNamespace.

But in the scenario where we have the unmanaged app using only the VRG(Hub less deployment), VRG can either be in the RamenOperandsNamespace OR the RamenOperatorNamespace.

This is only for the backward compatibility with the hub less mode. Let me know how to best document it.

BenamarMk commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -1284,3 +1288,101 @@ func removeString(values []string, s string) []string {
+	for _, vrg := range vrgs.Items {
+		log1 := log.WithValues("vrg", vrg.Name)
+
+		if vrg.Spec.ProtectedNamespaces == nil {
```
`if vrg.Spec.ProtectedNamespaces == nil || len(vrg.Spec.ProtectedNamespaces) == 0 {`

raghavendra-talur replied:
done

ShyamsundarR commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,147 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
+	common, err := r.drpcHaveCommonClusters(ctx, drpc, otherDRPC, log)
+	if err != nil {
+		return fmt.Errorf("failed to check if drpcs have common clusters (%w)", err)
+	} else if !common {
```
(nit) style remove the else, just add another if below

raghavendra-talur replied:
done

ShyamsundarR commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,147 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
+		return nil
+	}
+
+	drpcIsInAdminNamespace := drpcInAdminNamespace(drpc, ramenConfig)
```
(nit) efficiency, check this first, than fetching clusters and policy for DRPCs etc. and return early

raghavendra-talur replied:
done

ShyamsundarR commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,147 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
 	return vrg
 }
+
+func drpcInAdminNamespace(drpc *rmn.DRPlacementControl, ramenConfig *rmn.RamenConfig) bool {
```
(nit) function order in our code is usually caller first and definition to follow caller. IOW we should have had `ensureDRPCValidNamespace` before `drpcInAdminNamespace` and so on ...

raghavendra-talur replied:
done

ShyamsundarR reviewed (commented):
NEed to review VRG changes, reviewed other changes till then.

ShyamsundarR commented on api/v1alpha1/drplacementcontrol_types.go:
```diff
@@ -132,6 +141,13 @@ type DRPlacementControlSpec struct {
+
+	// Label selector to identify all the kube objects that need DR protection.
+	// It will be passed in to the VRG when it is created
+	// +optional
```
optional here with the CEL rules allow an inital YAML that does not contain the field and then later a YAML apply that contains the field, based on the lastAppliedConfiguration (or so).

The pattern that we potentially want is to make this required, such that initial versions themselves set it or mark it empty.

raghavendra-talur commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,147 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
 }
+
+func drpcInAdminNamespace(drpc *rmn.DRPlacementControl, ramenConfig *rmn.RamenConfig) bool {
+	drpcAdminNamespaceNames := drpcAdminNamespaceNames(*ramenConfig)
```
done

raghavendra-talur commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,147 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
+		return nil
+	}
+
+	if drpc.Spec.ProtectedNamespaces != nil {
```
done

raghavendra-talur commented on api/v1alpha1/volumereplicationgroup_types.go:
```diff
@@ -185,6 +185,14 @@ type VolumeReplicationGroupSpec struct {
 	KubeObjectProtection *KubeObjectProtectionSpec `json:"kubeObjectProtection,omitempty"`
+
+	// ProtectedNamespaces is a list of namespaces that are considered for protection by the VRG.
+	// Omitting this field means resources are only protected in the namespace controlled by the PlacementRef.
```
done

raghavendra-talur commented on api/v1alpha1/ramenconfig_types.go:
```diff
@@ -155,6 +155,9 @@ type RamenConfig struct {
 	VolumeUnprotectionEnabled bool `json:"volumeUnprotectionEnabled,omitempty"`
+
+	// RamenOpsNamespace is the namespace where resources for unmanaged apps are created
+	RamenOpsNamespace string `json:"ramenOpsNamespace,omitempty"`
```
Gowtham pointed out that we should prevent change of this value once it is set.

BenamarMk replied:
It is somewhat restrictive, but I understand why Gowtham is concerned about changing it once it is been set. However, if the user were to delete all the unmanaged apps, wouldn't that provide an opportunity for the user to make changes?

ShyamsundarR replied:
Also, this would be an admin change, i.e the CM would be editable for an admin persona, and if they change it from a->b, DRPCs in a will report errors. I am sort of fine with that, cleanup `a` and move to NS `b`.

The alternatives would put some stress on the code which is better avoided.

But good observation though.

BenamarMk commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -694,6 +696,8 @@ func (r *DRPlacementControlReconciler) Reconcile(ctx context.Context, req ctrl.R
 
 	drpc := &rmn.DRPlacementControl{}
 
+	var placementObj client.Object
```
(nit) I would move this declaration to line 726 and pass a `nil` in line 719.  I prefer to be explicit when we have not attempted to assign a value to it.

raghavendra-talur replied:
Done.

ShyamsundarR commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -710,7 +714,17 @@ func (r *DRPlacementControlReconciler) Reconcile(ctx context.Context, req ctrl.R
-	placementObj, err := getPlacementOrPlacementRule(ctx, r.Client, drpc, logger)
+	_, ramenConfig, err := ConfigMapGet(ctx, r.APIReader)
+	if err != nil {
+		statusErr := r.updateDRPCStatus(ctx, drpc, placementObj, logger)
```
Shouldn't we be using recordFailure here?

raghavendra-talur replied:
Done.

ShyamsundarR reviewed (commented):
Broadly LGTM!

BenamarMk commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,152 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
+			return fmt.Errorf("drpc in admin namespace must have protected namespaces")
+		}
+
+		drpcAdminNamespaceNames := drpcAdminNamespaceNames(*ramenConfig)
```
(nit) Here it is the same variable name as the function name.

BenamarMk reviewed (commented):
Done with the review.

ShyamsundarR commented on controllers/vrg_pvc_selector.go:
```diff
@@ -14,29 +13,33 @@ type PvcSelector struct {
+// pvcNamespaceNamesDefault returns the default pvc namespaces for the VRG.
+// If the VRG namespace is the Ramen operands namespace, then the protected namespaces are used.
+// In the else cases, vrg in application namespace or the ramen operator namespace, the VRG namespace is used.
+func pvcNamespaceNamesDefault(vrg ramen.VolumeReplicationGroup, ramenConfig ramen.RamenConfig) []string {
```
This is a slight silent ignore of `ProtectedNamespaces` if set on the VRG. E.g a user namespace VRG (not created by DRPC), lists a set of NS then this function would silently ignore it than error out.

I was thinking if we could validate early regarding this mismatch and report an error, it would make it cleaner to process mentally.

ShyamsundarR commented on controllers/vrg_recipe.go:
```diff
@@ -66,20 +74,20 @@ func RecipeElementsGet(ctx context.Context, reader client.Reader, vrg ramen.Volu
+			CaptureWorkflow: captureWorkflowDefault(vrg, ramenConfig),
 			RecoverWorkflow: recoverWorkflowDefault(),
 		}
 
```
Comment on old code below, so a Recipe can be fetched from any namespace, that is neither the VRG NS or in the list of `ProtectedNamespaces`. Now, not sure that is intentional, hence starting a comment on the same.

raghavendra-talur commented on controllers/drplacementcontrol_controller.go:
```diff
@@ -2643,3 +2668,152 @@ func constructVRGFromView(viewVRG *rmn.VolumeReplicationGroup) *rmn.VolumeReplic
+		return []string{}, err
+	}
+
+	protectedNamespaces, err := selectVRGNamespace(r.Client, log, drpc, placementObj)
```
Done.

raghavendra-talur commented on controllers/ramenconfig.go:
```diff
@@ -236,12 +236,25 @@ func ConfigMapGet(
+	return []string{RamenOperandsNamespace(config), RamenOperatorNamespace()}
+}
+
+// drpcAdminNamespaceNames returns the namespace names where the drpc objects can be created for multi namespace protection. The list does not include the namespace where the ramen operator pod is running. The DRPC must be created only in RamenOperandsNamespace for multi namespace protection.
```
Done.

raghavendra-talur commented on controllers/ramenconfig.go:
```diff
@@ -236,12 +236,25 @@ func ConfigMapGet(
+}
+
+// drpcAdminNamespaceNames returns the namespace names where the drpc objects can be created for multi namespace protection. The list does not include the namespace where the ramen operator pod is running. The DRPC must be created only in RamenOperandsNamespace for multi namespace protection.
+func drpcAdminNamespaceNames(config ramendrv1alpha1.RamenConfig) []string {
```
Done.

ShyamsundarR commented:
Generic comment on existing code. `ListPVCsByPVCSelector` lists all PVCs in the cluster (using required label selector) and then filters out those that are in desired namespaces. Seems a lot, instead maybe better to iterate over namespaces from the list of protected namespaces.

ShyamsundarR commented:
I would prefer a latch in processVRG that determines that VRG will honor ProtectedNamespaces or not. This can help exit early and also only support the feature explicitly when VRG is in the ops (or ramen) namespace.

ShyamsundarR commented:
Some broad comments noted while reviewing (some are repeats form other discussions), just building a TODO for the future: (@raghavendra-talur and @BenamarMk add any more that you have, jut to keep track of the same)

- DRPC in user namespace can have k8sResource protection via Recipe or the selector, this seems to work as desired with the changes, but just noting it as a possible use-case for testing
- Should de let the first DRPC that "claimed" a NS to continue on conflicts if another DRPC attempts to protect the same NS?
  - For example, let DRPC status carry NS that it claims, so even on conflict if it has claimed an NS it will continue reconcile of its instance, whereas the DRPC that changed to add this NS would not
  - NOTE: Namespaces are not present on the hub to "claim" then using any annotation on the actual NS (maybe on the MC then?)
- RD/RS ownership
- k8s resource status and metrics
- Operand NS creation on MC
- If a protected NS is removed from a DRPC, should unprotect resources in those NS on the MC
  - Next backup would not consider the NS
  - For PVCs etc. a broader list and see who the owner is and unprotect scheme would continue to work I assume
  - DRPC still needs changes to support unprotect

ShyamsundarR commented:
> > Some broad comments noted while reviewing (some are repeats form other discussions), just building a TODO for the future: (@raghavendra-talur and @BenamarMk add any more that you have, jut to keep track of the same)
> > ```
> > * DRPC in user namespace can have k8sResource protection via Recipe or the selector, this seems to work as desired with the changes, but just noting it as a possible use-case for testing
> > 
> > * Should de let the first DRPC that "claimed" a NS to continue on conflicts if another DRPC attempts to protect the same NS?
> >   
> >   * For example, let DRPC status carry NS that it claims, so even on conflict if it has claimed an NS it will continue reconcile of its instance, whereas the DRPC that changed to add this NS would not
> >   * NOTE: Namespaces are not present on the hub to "claim" then using any annotation on the actual NS (maybe on the MC then?)
> > 
> > * RD/RS ownership
> > 
> > * k8s resource status and metrics
> > 
> > * Operand NS creation on MC
> > 
> > * If a protected NS is removed from a DRPC, should unprotect resources in those NS on the MC
> >   
> >   * Next backup would not consider the NS
> >   * For PVCs etc. a broader list and see who the owner is and unprotect scheme would continue to work I assume
> >   * DRPC still needs changes to support unprotect
> > ```
> 
>     * Add owner labels in `updateVR` to "upgrade" existing VRs with the owner label (not required as yet as backward compat is handled)

* Protected App has resources on peer cluster namespace? Should we ensure peer is clean of PVCs and other resources?
* k8s restore as secondary happens [each time VRG generation changes](), which will happen in the case of Volsync and hence a restore will be initiated
* [Talur] App NS creation on MC during DR actions, so that we have the NS available for PVC restores
* Automate cleanup when working with imperative applications

raghavendra-talur commented:
@ShyamsundarR @BenamarMk Addressed two critical issues:
1. moved the kubeObjectSelector to KubeObjectProtectionSpec
2. added a check in processVRG to ensure that a vrg that has protectedNamespaces set is in the ramen operands namespace.

This PR is ready for merge.

BenamarMk reviewed (approved):
It is fine as is. We can merge it.
